pluginconfig/execute.go

```diff
@@ -78,6 +81,128 @@ func (e ExecuteOffchainConfig) Validate() error {
+		if _, exists := set[key]; exists {
+			return errors.New("duplicate token data observer type and version")
+		}
+		set[key] = struct{}{}
+	}
+	return nil
+}
+
+// TokenDataObserverConfig is the base struct for token data observers. Every token data observer
+// has to define its type and version. The type and version is used to determine which observer's
+// implementation to use. Whenever you want to add a new observer type, you need to add a new struct and embed that
+// in the TokenDataObserverConfig similarly to how USDCCCTPObserverConfig is embedded in the TokenDataObserverConfig.
+// There are two additional checks for the TokenDataObserverConfig to enforce that it's semantically (Validate)
+// and syntactically correct (WellFormed).
+type TokenDataObserverConfig struct {
+	// Type is the type of the token data observer. You can think of different token data observers as different
+	// strategies for processing token data. For example, you can have a token data observer for USDC tokens using CCTP
+	// and different one for processing LINK token.
+	Type string `json:"type"`
+	// Version is the version of the TokenObserverConfig and the matching Observer implementation for that config.
+	// This is used to determine which version of the observer to use. Right now, we only have one version
+	// of the observer, but in the future, we might have multiple versions.
+	// This is a precautionary measure to ensure that we can upgrade the observer without breaking the existing ones.
+	// Example would be CCTPv1 using AttestationAPI and CCTPv2 using a different API or completely
+	// different strategy which requires different configuration and implementation during Observation phase.
+	// [
+	//  {
+	//    "type": "usdc-cctp",
+	//    "version": "1.0",
+	//    "attestationAPI": "http://circle.com/attestation",
+	//    "attestationAPITimeout": "1s",
+	//    "attestationAPIIntervalMilliseconds": "500ms"
+	//  },
+	//  {
+	//    "type": "usdc-cctp",
+	//    "version": "2.0",
+	//    "customCirlceAPI": "http://cirle.com/gohere",
+	//    "yetAnotherAPI": "http://cirle.com/anotherone",
+	//    "customCircleAPITimeout": "1s",
+	//    "yetAnotherAPITimeout": "500ms"
+	//  }
+	//]
+	// Having version in that JSONisn't expensive, but it could reduce the risk of breaking the observers in the future.
+	Version string `json:"version"`
+
+	*USDCCCTPObserverConfig
+}
+
+// WellFormed checks that the observer's config is syntactically correct - proper struct is initialized based on type
+func (t TokenDataObserverConfig) WellFormed() error {
+	switch t.Type {
+	case USDCCCTPHandlerType:
+		if t.USDCCCTPObserverConfig == nil {
+			return errors.New("USDCCCTPObserverConfig is empty")
+		}
+	default:
+		return errors.New("unknown token data observer type")
+	}
+	return nil
+}
+
+// Validate checks that the observer's config is semantically correct - fields are set correctly
+// depending on the config's type
+func (t TokenDataObserverConfig) Validate() error {
```

Comment:
Any reason not to combine this with `WellFormed`?

Reply:
I wanted to follow the existing pattern for the configs. Validate is a semantical check - verifying fields values. However, you can think of WellFormed as a syntactical check called during deserialization. We would not need that in an ideal world, because the JSON unmarshaller would cover it.

Reply:
Assuming multiple embedded types in the future, the json Unmarshaller would still allow multiple of them to be initialized.

Reply:
We only verify here if someone didn't create a JSON monster that contains all possible structs or something with mixed fields

Reply:
> Assuming multiple embedded types in the future, the json Unmarshaller would still allow multiple of them to be initialized.

We do have `DecodeExecuteOffchainConfig(encodedExecuteOffchainConfig []byte) (ExecuteOffchainConfig, error)` which is an entry point for deserialization and it's responsible for the JSON unmarshalling. This is also the place in which I call `WellFormed()` function. We can make this logic part of `Validate` but I feel it's better suited for `DecodeExecuteOffchainConfig` where we verify JSON against the expected schema. Therefore `DecodeExecuteOffchainConfig` always returns a syntactically valid object, meaning it would return an error if you try creating a single  `TokenDataObserverConfig` that contains configs for both USDC and MyFancyToken